### PR TITLE
ANSI Cursor Position Codes

### LINKED ISSUE #98: Update putchar to handle ANSI cursor position
**Is your feature request related to a problem?**
We should not have a function to print at a given location

**Describe the feature request in detail**
putchar (putc) should handle the ANSI code to move the cursor

**Additional context**
```
Esc[Line;ColumnH
Esc[Line;Columnf
```
**Cursor Position:**
Moves the cursor to the specified position (coordinates).
If you do not specify a position, the cursor moves to the home position at the upper-left corner of the screen (line 0, column 0). This escape sequence works the same way as the following Cursor Position escape sequence.

### PR DESCRIPTION
**Pull request by Kfeavel**
Closes #98 
TODO:

- [x] Save cursor position code
- [x] Restore cursor position code

### CODE REVIEW

File: kernel/dev/tty/tty.cpp [hunk 1]
```diff
@@ -96,31 +98,40 @@ static size_t ansi_values_index = 0;
...
 #define POP_VAL() ansi_values[--ansi_values_index]
 
-void putchar(char c) {
+int putchar(char c) {
     switch (ansi_state) {
         case Normal: // print the character out normally unless it's an ESC
             if (c != ESC) break;
             ansi_state = Esc;
-            return;
+            return c;
```
Comment (micahswitzer) [inline kernel/dev/tty/tty.cpp]: This may be a good candidate for `goto end` instead of many `return c`s

File: kernel/dev/tty/tty.cpp [hunk 2]
```diff
@@ -142,11 +153,36 @@ void putchar(char c) {
...
                 ansi_val = 0;
+            } else if (c == 'H' || c == 'f') { // Set cursor position attribute
+                PUSH_VAL(ansi_val);
+                // take action here
+                // iterate through all values
+                // the proper order is 'line (y);column (x)'
+                while (ansi_values_index > 0) {
+                    ansi_val = POP_VAL();
+                    // If we only have one item in the stack now (i.e. line is done)
+                    if (ansi_values_index == 1) {
```
Comment (micahswitzer) [inline kernel/dev/tty/tty.cpp]: What happens if there were more or less than 2 values provided? Should it always accept them and just only look at the first two pushed?

File: kernel/dev/tty/tty.cpp [hunk 3]
```diff
@@ -188,4 +224,21 @@ void putchar(char c) {
...
         tty_coords_y = X86_TTY_HEIGHT - 1;
     }
+    return c;
+}
+
+int puts(const char *str) {
+    int i = 0;
+    // Loops until a null character
+    while(str[i]) {
+        if(putchar(str[i]) == EOF) {
```
Comment (micahswitzer) [inline kernel/dev/tty/tty.cpp:236]: Maybe you just haven't finished writing the code yet, but I don't see any code in `putchar` that returns `EOF`.

File: kernel/main.cpp [hunk 4]
```diff
@@ -118,22 +118,24 @@ extern "C" void px_kernel_main(const multiboot_info_t* mb_struct, uint32_t mb_ma
...
     int i = 0;
     while (true) {
         // Display a spinner to know that we're still running.
+        // Each line will save the cursor position, change it
+        // to the top right corner (-1) and then restore the
+        // cursor position (to the saved location).
         switch (i) {
             case 0:
-                px_kprintf("\b|");
+                px_kprintf("\033[s\033[%i;%iH\b|\033[u", X86_IND_Y, X86_IND_X);
```
Comment (micahswitzer) [inline kernel/main.cpp]: I understand wanting to test with this, but maybe it should be merged in separately once this PR is ready?

File: Makefile [hunk 5]
```diff
@@ -218,11 +218,11 @@ vbox: vbox-create
 
 # Open the connection to qemu and load our kernel-object file with symbols
 .PHONY: debug
-debug: dist/panix.iso
+debug: dist/panix.kernel
```
Comment (micahswitzer) [inline Makefile]: These changes are unrelated to this PR

File: kernel/dev/tty/tty.cpp (hunk 2 above)
Comment (Kfeavel) [reply kernel/dev/tty/tty.cpp]: > What happens if there were more or less than 2 values provided? Should it always accept them and just only look at the first two pushed?

I think that would be easiest. It would be non-standard to provide more than 2 values, so the safest thing to do is just ignore them. The code will need a little tweaking to check for that, but it won't be too hard.

File: kernel/dev/tty/tty.cpp (hunk 1 above)
Comment (Kfeavel) [reply kernel/dev/tty/tty.cpp]: That's fair. Is end already a label or will I need to update all of the returns to `goto end`?

File: kernel/dev/tty/tty.cpp (hunk 3 above)
Comment (Kfeavel) [reply kernel/dev/tty/tty.cpp:236]: In what cases should putchar return EOF? putchar is supposed to return the character printed, so do we return EOF on null / `\n`, or in what cases?

File: kernel/main.cpp (hunk 4 above)
Comment (Kfeavel) [reply kernel/main.cpp]: Fair enough. I'll remove it and the makefile change.

**Comment by Kfeavel** [comment]
At this point I'm pretty much ready to merge except I need @micahswitzer to explain how he would like the `goto` code to be implemented (since he already has an idea in mind) and in what cases `EOF` should be returned.

File: kernel/dev/tty/tty.cpp (hunk 1 above)
Comment (micahswitzer) [reply kernel/dev/tty/tty.cpp]: You'll need to add it.

File: kernel/dev/tty/tty.cpp (hunk 3 above)
Comment (micahswitzer) [reply kernel/dev/tty/tty.cpp:236]: I have no idea, I saw that you were depending on `putchar` returning `EOF` so I was clarifying what the expected behavior should be.

File: kernel/dev/tty/tty.cpp (hunk 2 above)
Comment (micahswitzer) [reply kernel/dev/tty/tty.cpp]: It looks like it will already behave that way
